src/python/pants/backend/python/packaging/pyoxidizer/rules.py

```diff
@@ -185,8 +186,17 @@ async def package_pyoxidizer_binary(
     )
 
 
+@rule
+async def run_pyoxidizer_binary(field_set: PyOxidizerFieldSet) -> RunRequest:
+    binary = await Get(BuiltPackage, PackageFieldSet, field_set)
+    artifact_relpath = binary.artifacts[0].relpath
+    assert artifact_relpath is not None
```

Comment:
Wait actually this won't consistently work because the code for `package` means you might have >1 `binary.artifacts`, unlike Go where we only ever have a single file. 

I don't know much about the non-single-file instances, other than your blog. Thoughts on a heuristic to identify what is the binary the user should execute? Maybe looking at the `config.rendered()` to figure out the binary name...? Then looking up the `install/` folder?

Reply:
Hmm, from my testing, I had assumed that the primary artifact was the named one that we specify, and everything else is later in the array. We know the name of the file it will generate, since that is specified - and I guess it was a coincidence in my testing that I always hit the "correct", top level item in install. 

In the generated config, we're specifying the name, but this may not hold for the templated versions

```python
 exe = dist.to_python_executable(
        name="$NAME",
        packaging_policy=policy,
        config=python_config,
    )
```

Reply:
And even more now that I think about it - this could roll back to the `output_path` stuff as well with how the file is named.

Reply:
> Hmm, from my testing, I had assumed that the primary artifact was the named one that we specify, and everything else is later in the array. 

Yeah that isn't reliable -- the artifacts are ordered by the `Snapshot.files` ordering, which is alphabetical. 

> And even more now that I think about it - this could roll back to the output_path stuff as well with how the file is named.

How so? `output_path` shouldn't impact how the file itself is named, only the prefix to the PyOxidizer portion of the path. We don't touch any of PyOxy's stuff other than stripping `build/`.

Reply:
I didn't mean that changing the output path would affect it, but kinda the other way around - if we want `output_path` to set the name for the target.

Not a big deal today, just started thinking about it.

Under what conditions can the artifact's `relpath` be none? Because shouldn't something like this work?

```python
artifact = next(artifact for artifact in binary.artifacts if artifact.relpath.endswith(field_set.address.target_name))
```

Reply:
If this is reasonable, I'll flip this out of draft

Reply:
>  I modified this to specifically look for the target_name with an install parent - so that should be pretty specific.

This works great, so long as the user is not using a custom config that sets the `NAME` to something else. Two thoughts on how we could reduce this risk:

1. Add to the `help` message of `template` that you should consume `$NAME`, otherwise `{bin_name()} run` won't work, or
2. Can we be confident that `install/` only has a single file? That's a good find about `install/lib`! 

>  and then only multiple targets from the same project can land in the same folder (and I think the changes to the default output_path even preclude that from happening, so the user would have to manually set the output_path of multiple targets to the same location).

This isn't possible. This is what I meant by the above comment that we're only inspecting the `Digest` returned by the `Package` goal; we are not inspecting `dist/` folder on-disk. There's no need to worry about multiple `pyoxidizer_binary` targets, we're guaranteed to only be dealing with the result of a single target.

So, so long as a single `pyoxidizer_binary` can only result in one file in the `install/` folder (ignoring `install/lib`), then we could simply use that single file. Meaning we work with the `template` field.

Fwit, we could add an `assert(len(matching_files)) == 1`, and ask users to file a bug report if the assertion is not valid the wild. Totally reasonable for an experimental backend.

Reply:
> This isn't possible. This is what I meant by the above comment that we're only inspecting the `Digest` returned by the `Package` goal; we are not inspecting `dist/` folder on-disk. There's no need to worry about multiple `pyoxidizer_binary` targets, we're guaranteed to only be dealing with the result of a single target.

Ahh, okay okay - I always kinda wondered what was happening here. Was never really sure which was being done under the hood.

> Fwit, we could add an assert(len(matching_files)) == 1, and ask users to file a bug report if the assertion is not valid the wild. Totally reasonable for an experimental backend.

I think this is a reasonable approach - this would prevent users from calling the `run` command if its not accounted for, but might be quick to fix either way. 

To the best of my knowledge, based on what I've seen after your modifications to the output files, there should only be 1 file. There used to be multiple files - only because PyOx builds to the same directory. Since you now isolate by `target_name`, that precludes multiple files. 

By default, PyOx is supposed to generate single-files, and the libs are actually a side-effect of not being able to put everything in memory. So, I can't see a use case where multiple binaries in 1 folder is *expected*. I haven't seen a use case for that in the docs I've read so far.

I'll filter out files in the `install` directory, and assert on that single file explicitly and ignore `target_name` entirely.

Won't be able to do this until tonight or tomorrow, I'm wiping and re-installing my OS, so the rest of the day will probably be getting setup again.

Reply:
> this would prevent users from calling the run command if its not accounted for, but might be quick to fix either way.

Agreed. FYI this is an example assertion we have:

```
            raise AssertionError(
                "A module has both unambiguous and ambiguous owners, which is a bug in the "
                "dependency inference code. Please file a bug report at "
                "https://github.com/pantsbuild/pants/issues/new."
            )
```

> I'll filter out files in the install directory, and assert on that single file explicitly and ignore target_name entirely.

Great! Sounds good. 

> Won't be able to do this until tonight or tomorrow, I'm wiping and re-installing my OS, so the rest of the day will probably be getting setup again.

Oof good luck 😅